### PR TITLE
Fix memory leak in netty connection pool

### LINKED ISSUE #2906: Regression - Netty Connection Exceptions with Large Connection Pool Count
Using snapshot: [3.0.0-RC8+22-5ea81ef6-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/dev/zio/zio-http_2.13/3.0.0-RC8+22-5ea81ef6-SNAPSHOT/)

There is a significant regression in request serving when using both `ConnectionPoolConfig.Dynamic` and `ConnectionPoolConfig.Fixed`. This seems to be a regression as this issue only appeared after migrating from RC6 to RC8. Exceptions include: 

```
[info] Exception in thread "zio-fiber-823178100,1568885665" io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection reset by peer: /0.0.0.0:8080
```
```
[error] Jun 13, 2024 3:25:18 PM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException Jun 13, 2024 3:25:18 PM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
```
```
Exception in thread "zio-fiber-1756387034" io.netty.handler.codec.PrematureChannelClosureException: Channel closed while executing the request. This is likely caused due to a client connection misconfiguration
```

Reproducer:
```scala
package repro

import zio.{ Config => _, *}
import zio.http.*
import zio.http.ZClient.Config
import zio.http.netty.NettyConfig

object Reproducer extends ZIOAppDefault {
  private val clientLayer =
    (ZLayer.succeed(Config.default.copy(connectionPool = ConnectionPoolConfig.Dynamic(1, 512, 60.second))) ++ ZLayer
      .succeed(NettyConfig.defaultWithFastShutdown) ++
      DnsResolver.default) >>> Client.live

  private val url = URL.decode("http://0.0.0.0:8080/Test").toOption.get
  private val req = Request.post(url, Body.empty)

  private val routes =
    HttpApp(
      Routes(
        Method.POST / "Test" ->
          handler(ZIO.sleep(1.second).as(Response.text("success")))
      )
    )

  val run: ZIO[Any & ZIOAppArgs & Scope, Throwable, Unit] = {
    for {
      _ <- Server.serve(routes).fork
      client <- ZIO.service[Client]
      _ <- ZIO.foreachParDiscard(0 to 400) { _ =>
        ZIO
          .scoped[Any] {
            client.request(req).flatMap(_.body.asArray)
          }
          .repeatN(100)
      }
    } yield ()
  }.provide(clientLayer, Server.default)
}
```

### PR DESCRIPTION
**Pull request by kyri-petrou**
/fixes #2906

This is totally on me, big apologies on this one. In #2843 I changed the logic to use a listener on Netty's close future instead of forking a fiber to await on the future. However, I was not removing the listener when the channel would be reset. This is causing the `onResult` promise to being referenced and doesn't allow the response to be GC'd, causing a rather severe memory leak until the channel is closed.

In addition, I found that one of the main things that was contributing massively to the memory leak was using a much-bigger-than-needed ZStream buffer in `AsyncBody`. I added dynamic sizing of the buffer when the content length of the response is known. This way we reduce memory allocations significantly for small requests.

Edit:
1. ~~After feedback from @Not-Aru, I reverted the implementation of listening to the closeFuture back to the one used in versions <= 3.0.0-RC6~~
2. I split the logic for handling http / websocket requests `NettyClientDriver#requestOnChannel`. It was making things difficult to reason with plus we were executing effects that were only meant for running websockets when handling http requests and vice-versa. I think the new implementation is much cleaner and easier to follow

Edit 2:
1. After some more extensive testing, I found a small race condition when interrupting the channel and fixed it
2. Using a FutureListener is much more performant than forking a scoped effect to monitor the closeFuture so I changed the implementation back

### CODE REVIEW

**Comment by hearnadam** [comment]
>  Continuous Integration / Publish Artifacts (pull_request) Skipped 

Can this be run? I would like to test the fix

**Comment by 987Nabil** [comment]
@hearnadam If you want to test this locally, please check out the branch and do a `publishLocal`.
Or wait for the snapshot after a merge

**Comment by Not-Aru** [comment]
We did a `publishLocal` and saw an improvement when testing performance over RC8, but not on par with RC6. It is unclear if this is because of our implementation or if this commit still does not completely fix the issue. I will deep dive into the logging to investigate further.

**Comment by kyri-petrou** [comment]
@Not-Aru I'll push another commit, let me know if things are improved.

PS: I identified some optimizations for the client, but I want to make them in a separate PR once this one has been merged in. Based on some local testing it improves performance by 15-20% but would really appreciate if you could help me with any benchmarking. I can tag you in the PR once I open it

**Comment by kyri-petrou** [comment]
@Not-Aru you can pull the latest changes now and `publishLocal` again

**Comment by kyri-petrou** [comment]
@hearnadam @Not-Aru if you don't mind, could you test [this branch](https://github.com/kyri-petrou/zio-http/tree/improve-client-non-streaming-performance) against your benchmarks?

It improves the Client performance quite substantially for responses we read as asArray / asChunk / asString etc. According to some local testing it's ~20-25% more performant than the current implementation. I'll open a PR with those changes once this one is merged in but want to check with you if you're not seeing some other regression from those changes

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
Merged right away due to severity. 👍

**Comment by russwyte** [comment]
Hey folks - it seems CI is broken and this snapshot is not publishing. Also can we expect a release since this is an important bug fix?

**Pull request merged**
